### ISSUE TITLE
usolve misses solutions

### ISSUE THREAD

**Issue by m93a** [josdejong/mathjs#1748]
The [`usolve`](https://mathjs.org/docs/reference/functions/usolve.html) function can only find solutions of non-degenerate matrices. On degenerate ones, it automatically sets the free parameters to zero, only returning one solution out of many. For null `b` it only returns null vectors as a result. Example:

```javascript
M = [[-2,3,5],[0,-1,2],[0,0,0]]
b = [0,0,0]

math.usolve(M, b)
// returns [[0], [0], [0]]
// correct result [5.5, 2, 1]
```

(Also, the returned array is two-dimensional for some reason.)

**EDIT**: Related to #1168. Since all non-square matrices are degenerate.
**EDIT**: Also true for [`lsolve`](https://mathjs.org/docs/reference/functions/lsolve.html) and [`lusolve`](https://mathjs.org/docs/reference/functions/lusolve.html).

**Comment by m93a** [comment]
I couldn't find a general back-substituion algorithm for singular matrices, so I spent some time figuring it out. I wrote [this paper](https://www.overleaf.com/read/csvgqdxggyjv) explaining why I think my algorithm is correct.

**Comment by josdejong** [comment]
Thanks for sharing your work Michal!

**Comment by josdejong** [comment]
Published now in `v7.3.0` 🎉
